### ISSUE TITLE
Notepad++ still doesn't handle scaling correctly

### ISSUE THREAD

**Issue by hapklaar** [notepad-plus-plus/notepad-plus-plus#6284]
### Description of the Issue
When opening Notepad++ on 1440p monitor and then dragging the window to a 4k (2160p) monitor with 150% scaling enabled, text is only zoomed in and becomes blurry.

Apps that are fully dpi/scaling aware don't show this and change font to a more suitable one. For example MS Outlook.

Issue does not occur when main monitor is the 4k monitor, only when main monitor is of lower resolution or scaling.

### Steps to Reproduce the Issue

1. Open Notepad++ with main screen 1440p
2. Drag Notepad++ window to secondary 4k monitor
3. Observe blurry text.

### Expected Behavior
Expected font to change to higher DPI one and get same result as opening Notepad++ when the 4k monitor is configured as main display. We can see Notepad++ can actually do this, because it works correctly.

### Actual Behavior
Blurry text on 4k monitor after dragging window from 1440p to 2160p monitor with 150% scaling enabled.

### Debug Information
Screenshots:
Normal 1440p layout:
![image](https://user-images.githubusercontent.com/7703320/67789916-1ca1e380-fa75-11e9-963b-4c55c63017f2.png)

After dragging screen to 4k monitor (blurry text):
![image](https://user-images.githubusercontent.com/7703320/67789973-32170d80-fa75-11e9-902a-57be58017462.png)

Opening notepad++ on 4k monitor with this 4k monitor as main display in Windows (shart and crispy text):
![image](https://user-images.githubusercontent.com/7703320/67790116-7efae400-fa75-11e9-9a0a-3a5c8f826b12.png)

**Comment by luigip** [comment]
See also #7584 and #2871

**Comment by hapklaar** [comment]
Just checking in, will this ever be picked up by a dev?

**Comment by krisztian-jani** [comment]
The issue is still live.

**Comment by ljoseph21** [comment]
Hope this helps.
https://support.microsoft.com/en-us/windows/fix-apps-that-appear-blurry-in-windows-10-e9fe34ab-e7e7-bc6f-6695-cb169b51de0f

**Comment by sasumner** [comment]
@ljoseph21 and others:
There will be a ["fix"](https://github.com/notepad-plus-plus/notepad-plus-plus/commit/1364b199b4eff3aa2bdc370e6adf6d6186576218) released in Notepad++ 7.9.3 which hopefully improves this.

**Comment by CBTMike** [comment]
> 
> 
> @ljoseph21 and others:
> There will be a ["fix"](https://github.com/notepad-plus-plus/notepad-plus-plus/commit/1364b199b4eff3aa2bdc370e6adf6d6186576218) released in Notepad++ 7.9.3 which hopefully improves this.

The problem is a lot worse in 7.9.3, at least for me.  :(

https://github.com/notepad-plus-plus/notepad-plus-plus/issues/9579

**Comment by sasumner** [comment]
>The problem is a lot worse in 7.9.3, at least for me. :(

Better for some, worse for others...it is ongoing work.

**Comment by luigip** [comment]
I just tried out the new version. My primary display (big external monitor) is at 100% scaling and my laptop display is at 125%.

The update just seems to have stopped the application scaling altogether, so the text and UI is at 100% on the laptop display too, i.e. too small.

Thanks for giving this some attention, but unless there's some setting I need to change, I can confirm that this fix has not worked.

**Comment by CBTMike** [comment]
> 
> 
> I observe the same as @luigip - the scaling on the secondary monitor is not taking place
> 
> Main monitor = UHD@125% scaling (28")
> Secondary monitor = UHD@225% scaling (labtop 14")
> 
> I ended up here searching for the a solution :-)

I've rolled back to 7.9.2.  From another thread it looks like they're rolling back the changes they made in display scaling in 7.9.3.  Hopefully that will fixed be in 7.9.4.  😄

**Comment by hapklaar** [comment]
Just updated to 7.9.5, but issue persists unchanged as far as I can see.

**Comment by CBTMike** [comment]
No issues for me since updating to 7.9.5.  Did not test 7.9.4.  7.9.3 was unusable on my 4K screen.  Hopefully they'll get this resolved for everyone else.  Good luck!

**Comment by MHillier98** [comment]
Same issue here with 7.9.5 - I have a 1440p display in a portrait (vertical) orientation, at 125% scaling.

**Comment by owencompher** [comment]
I had the same issue. My laptop display was at 125% scaling and so the text was a bit blurry. I was able to fix the issue by changing the program Properties>Compatibility>Change high DPI settings>High DPI scaling override>Scaling preformed by: System (Enhanced) 
![image](https://user-images.githubusercontent.com/72040115/123522033-69d1c080-d66f-11eb-9ae7-5bbd2f2cd0db.png)
This used the system scaling factor without causing the blurriness.

**Comment by N0xFF** [comment]
> I had the same issue. My laptop display was at 125% scaling and so the text was a bit blurry. I was able to fix the issue by changing the program Properties>Compatibility>Change high DPI settings>High DPI scaling override>Scaling preformed by: System (Enhanced)
> ![image](https://user-images.githubusercontent.com/72040115/123522033-69d1c080-d66f-11eb-9ae7-5bbd2f2cd0db.png)
> This used the system scaling factor without causing the blurriness.

After enabling, Notepad++ blurriness was fixed on standard DPI but became blurry on high DPI. Before, it was contrary.

**Comment by hapklaar** [comment]
More than 2 years after my initial post the issue still exists. Is there any chance in it ever getting fixed?

**Comment by unit-a-user** [comment]
Currently `System (Enhanced)` pixelated quality.

**Comment by eladkarako** [comment]
until fixed, there is a patch https://github.com/notepad-plus-plus/notepad-plus-plus/issues/10323#issuecomment-1114771445

**Comment by GravelX** [comment]
I'm still experiencing issues regarding scaling on second monitor. Just bumping this thread

**Comment by cloudgreen** [comment]
Loving Notepad++ but you get very weird display issues if you use 125% screen scaling. Fixing this would be fantastic! :)

**Comment by kukoman** [comment]
yep, still experiencing the issue as well, full hd laptop + 4k connected through DP, scaling 100%/125% and having blurred text as well :(

**Comment by JackAidley** [comment]
Still getting this problem. In my case 125% on my main 1440 monitor and 100% on the secondary 1080. Using the High DPI settings gets clear text on the second monitor but makes it blurry on the primary.

**Comment by iampradip** [comment]
> I had the same issue. My laptop display was at 125% scaling and so the text was a bit blurry. I was able to fix the issue by changing the program Properties>Compatibility>Change high DPI settings>High DPI scaling override>Scaling preformed by: System (Enhanced) ![image](https://user-images.githubusercontent.com/72040115/123522033-69d1c080-d66f-11eb-9ae7-5bbd2f2cd0db.png) This used the system scaling factor without causing the blurriness.

This fixes blurriness but the monospace font is not displayed properly i.e. when you move cursor up down in code you can see cursor going left-right in each line. Ideally it should stay in same position since monospace font has each character having same width.

**Comment by LECOQQ** [comment]
Still not resolved. The quick fix allows me to have npp on my second display without bluriness, but the main display is not blurrier than before.

**Comment by haja-fgabriel** [comment]
Still not solved in v8.4.8

**Comment by marcoalesan** [comment]
Still not fixed in v8.4.9 🫤

**Comment by benlangfield** [comment]
Was wondering what was causing it, if I open on my laptop then later dock the laptop and drag to my monitor which is a different resolution it causes it (as outlined by OP)

**Comment by il-mix** [comment]
Problem is still there with v.8.5.3.
At the moment Notepad++ DPI awareness is "System Aware" (default).
A solution could be to set it to "Per-monitor Aware"
https://learn.microsoft.com/en-us/windows/win32/api/shellscalingapi/nf-shellscalingapi-setprocessdpiawareness

**Comment by kukoman** [comment]
still not fixed in v8.5.4 :/

**Comment by hackgrid** [comment]
How is this still an issue? :(

**Comment by CBTMike** [comment]
I haven't had this issue in like two years.  Across many different monitors.  Standard 1920x1080 24" monitors, 34" and 38" curved screens, 4K screens, etc.  All work fine.  People who are still reporting a problem should maybe post a picture or screenshot or something.  I'm guessing it's "user error" or something off with their particular config.

**Comment by kukoman** [comment]
well, "it works for me" isn't really an answer... If you read the thread
carefully, you will see pictures and reports from dozens of people...

**Comment by hackgrid** [comment]
> I haven't had this issue in like two years. Across many different monitors. Standard 1920x1080 24" monitors, 34" and 38" curved screens, 4K screens, etc. All work fine. People who are still reporting a problem should maybe post a picture or screenshot or something. I'm guessing it's "user error" or something off with their particular config.

The problem appears, when you use different resolution monitors at once in a multi monitor setup.

e.g. i have 2 1080p displays + 1 2160p display, and notepad++ is always blurry!

The title bar is sharp, all text in notepad++ is blurry!

![npp-blurry](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/54060/10d4b4ef-c5f3-4a53-8963-db542f3240bd)

**Comment by iampradip** [comment]
> I haven't had this issue in like two years. Across many different monitors. Standard 1920x1080 24" monitors, 34" and 38" curved screens, 4K screens, etc. All work fine. People who are still reporting a problem should maybe post a picture or screenshot or something. I'm guessing it's "user error" or something off with their particular config.

Issue is easily reproducible when you use two monitors with different DPI settings, especially when both have non-100% DPI.

**Comment by haja-fgabriel** [comment]
Windows sends a WM_DPICHANGED message to Notepad++ when the DPI gets changed. Handling this message is not implemented yet. Please take a look at this:
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/7234e1478c36ccf2732301b419aa9ae03bbc0558/PowerEditor/src/NppBigSwitch.cpp#L3319

This could be another useful link:
https://learn.microsoft.com/en-us/windows/win32/hidpi/high-dpi-desktop-application-development-on-windows

**Comment by VVelda** [comment]
The editor becomes blurry also when one uses it through RDP (with different DPI), so this is extremely annoying. :-/
Where can we send donations _dedicated to fixing this_ (now almost eternal) bug?

**Comment by lucidnx** [comment]
I wonder why is this so hard to implement in almost every app...

**Comment by iampradip** [comment]
> I wonder why is this so hard to implement in almost every app...

It is extremely hard.

**Comment by Maelstromeous** [comment]
> Windows sends a WM_DPICHANGED message to Notepad++ when the DPI gets changed. Handling of this message is not implemented yet. Please take a look at this:
> 
> https://github.com/notepad-plus-plus/notepad-plus-plus/blob/7234e1478c36ccf2732301b419aa9ae03bbc0558/PowerEditor/src/NppBigSwitch.cpp#L3319
> 
> Another place where this message is handled is in the ScintillaWin class:
> 
> https://github.com/notepad-plus-plus/notepad-plus-plus/blob/7234e1478c36ccf2732301b419aa9ae03bbc0558/scintilla/win32/ScintillaWin.cxx#L2176
> 
> https://github.com/notepad-plus-plus/notepad-plus-plus/blob/7234e1478c36ccf2732301b419aa9ae03bbc0558/scintilla/win32/ScintillaWin.cxx#L2181
> 
> This could be another useful link: https://learn.microsoft.com/en-us/windows/win32/hidpi/high-dpi-desktop-application-development-on-windows#per-monitor-dpi-scaling-support-by-ui-framework--technology
> 
> Microsoft has also published a demo for per-monitor DPI awareness on their GitHub: https://github.com/Microsoft/Windows-classic-samples/tree/main/Samples/DPIAwarenessPerWindow
> 
> Issue #7584 also addresses this issue: [#7584 (comment)](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/7584#issuecomment-991793497)
> 
> Later edit: if I revert the changes from this commit, the breakpoint does stop inside the places, but I am not sure if the bugs are solved

The fix is hopefully here, devs can we take a look at this potential soluition please?

**Comment by xomx** [comment]
> The fix is hopefully here

https://github.com/notepad-plus-plus/notepad-plus-plus/issues/14337#issuecomment-1830828245

It is a lot of work.

**Comment by JackAidley** [comment]
It is fixed for me now.

**Comment by hackgrid** [comment]
8.6.0 still had the issue.
I upgrade to 8.6.2 and it is fixed! :-)
Also 8.6.4 works.

Really great, thank you!

I could not find the npp commit that fixed this, I guess the update to Scintilla 5.4.1 is responsible?

1080p display:
![grafik](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/54060/32c069a6-0cef-4420-a24f-12e5751a3738)

2160p display:
![grafik](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/54060/0328c2ed-f778-4419-8288-7775f2d7a6c9)

**Comment by JackAidley** [comment]
Yeah, looking closely, I think it is much improved but the lower res display is not showing as crisp as it should.

**Comment by mobeigi** [comment]
I just changed one of my monitors to a 4K one (coming from a dual 1440p setup).
I'm running 4K monitor ( 175% scaling) and a 1440p monitor (125% scaling).

On the 4k monitor its very crisp, on the 1440p monitor its slightly blurry / not as crisp.
If I change Windows 10 scaling on my 1440p monitor to match 175% scaling (like my main monitor) then it also becomes crisp.
Seems like the app is picking one DPI and sticking with it.

Man there is so much hassle in running a dual monitor setup using different monitors / resolutions etc. I really should save myself a headache and always use same monitors / resolutions in the future.

**Comment by hackgrid** [comment]
Today I discovered the NotepadNext reimplementation of Notepad++:

https://github.com/dail8859/NotepadNext

This alternative does not have the blurry issue on multi-monitor-with-different-dpi-settings at all!

So if you can live with the current shortcomings of NN compared to N++, this may be a solution for you all!

I doubt somebody could backport their solution to N++, because they did a complete rewrite using QT6.

**Comment by hackgrid** [comment]
5/72 false positives on Virustotal mean nothing! And Version 0.7 has 0 detections on Virtustotal right now, what file did you test??
Also this project is open-source too with a long history, nothing to be afraid about?

**Comment by hackgrid** [comment]
still some false positives based on heuristics from some no name scanners, nothing to worry about

**Comment by hackgrid** [comment]
McAfee is shit since 10 years +

**Comment by JonWillis-Tote** [comment]
> I just changed one of my monitors to a 4K one (coming from a dual 1440p setup). I'm running 4K monitor ( 175% scaling) and a 1440p monitor (125% scaling).
> 
> On the 4k monitor its very crisp, on the 1440p monitor its slightly blurry / not as crisp. If I change Windows 10 scaling on my 1440p monitor to match 175% scaling (like my main monitor) then it also becomes crisp. Seems like the app is picking one DPI and sticking with it.
> 
> Man there is so much hassle in running a dual monitor setup using different monitors / resolutions etc. I really should save myself a headache and always use same monitors / resolutions in the future.

I have this issue using 4k monitors with ultrawide in the middle. Its not just notepad, even TaskManager and VisualStudio have issues with it. Feels like something that Window should be able to handle, in that it renders the app for the DPI of each monitor, might mean rendering it twice, but its better than the blurry mess you get now.

**Comment by hackgrid** [comment]
Strange, the inbuilt Windows apps work fine for me!

On which Windows version are you?

From Windows 10, version 1607 on it works best:
https://learn.microsoft.com/en-us/windows/win32/hidpi/setting-the-default-dpi-awareness-for-a-process

**Comment by roelandschoukens** [comment]
This issue is definitely **not fixed** in Notepad++ 8.6.5.

Notepad++ is DPI-aware, but not per-monitor DPI aware. So it gets one DPI value when it starts up. *This DPI comes from your primary monitor*, so it is going to appear sharp on that monitor, and blurry on others.

If you use Scaling preformed by: System (Enhanced), it will let Notepad++ work as if you have 100% DPI scaling, but the underlying drawing APIs will render to a surface with twice the resolution (or maybe more), and then scale that to your actual monitor DPI. Text and path primitives will thus look sharp-ish. This seems to work reasonably well with Notepad++. But it is a workaround for programs that are not DPI aware at all and will not look as good as DPI scaling done by the program itself.

**Comment by photopea** [comment]
Let me explain the problem with an example.

I want a letter "F" to be 10 hardware pixels tall on my 15" 1080p laptop, but I want it to be 12 hardware pixels tall on my 30" 4K monitor (a hardware pixel is 0.18 x 0.18 mm on both displays, but I sit further from the 30" monitor).

Text is rendered nicely on my laptop (left image), but when I connect my monitor, the rendered text is just upscaled to 120% using the bilinear interpolation method (middle image). You need to restart Notepad++ to make it render text nicely (right image). 

Other applications (e.g. a web browser) can detect the change in the expected scale, and they re-render text without restarting the app.

![Image](https://github.com/user-attachments/assets/9c002721-74ce-44f3-aa14-2257f470c105)
